### PR TITLE
[DO NOT MERGE] Bank holidays location experiment

### PR DESCRIPTION
**Pull request by andysellick**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Add functionality to the bank holidays page to let users save their location (nation) so that the next time they visit, they get shown the relevant bank holiday tab automatically.

How it works:

- only visible if user has accepted cookies
- user can choose to save their location by clicking the button in the preferred tab, saves the location e.g. `Scotland` in a cookie
- when they return to the page `/bank-holidays` the tab they've chosen will be automatically shown
- on mobile, the screen jumps to the relevant part of the page (there aren't any tabs on mobile)

There's some complexity around the URL:

- if the URL has a fragment on it e.g. `/bank-holidays#scotland` the page ignores a previously saved location for that visit, as we assume users are intentionally trying to view that bank holiday
- instead, it will show the information that matches the fragment, in this case the Scotland bank holidays

## Why

This is a personalisation experiment. We want to see if users find this useful and how they interact with it. This will be run for a limited time and either taken down or improved after that.

## Visual changes

Screenshot 2021-03-31 at 11 02 59

and after saving:

Screenshot 2021-03-31 at 11 03 06

Trello card: https://trello.com/c/qm6BnJiO/680-public-holidays-location-experiment-fe

### CODE REVIEW

File: app/views/components/_subscribe.html.erb [hunk 1]
```diff
@@ -1,3 +1,3 @@
 <%= tag.div class: "app-c-subscribe" do %>
-  <%= link_to label, url, title: title, class: "app-c-subscribe__link govuk-link" %>
+  <%= link_to label, url, title: title, class: "app-c-subscribe__link govuk-link", data: data %>
```
Comment (danacotoran) [inline app/views/components/_subscribe.html.erb:2]: This `data` parameter is making the tests fail for the  clocks change page which [also uses this partial, apparently](https://github.com/alphagov/frontend/blob/85ddb3ef9ae62b359459861f61fdfe68df0c46f7/app/views/calendar/when_do_the_clocks_change.html.erb#L35).

**Review by danacotoran — COMMENTED** [review]
Ok just a few notes before I log off for the day (running this locally with Docker FWIW):

1. when I switch between tabs with VoiceOver, it reads out "Do you want to save X as your... etc" immediately after announcing the tab has been selected –  is this intended behaviour? https://user-images.githubusercontent.com/7116819/113029284-2527ad80-9184-11eb-8814-d1ec30144f39.mov
2. when I press the button to save a country, the state updates as expected, however when I try to undo my action, the button no longer seems to respond https://user-images.githubusercontent.com/7116819/113029687-9e270500-9184-11eb-985e-e290b6020f7e.mov
3. when I select a country as my default location, the UI state is updated but it is not maintained when I refresh the page. When I visit `/bank-holidays` without the hash, it shows me the default view. From poking around in dev tools, I can see it is not dropping the cookie as intended (which is probably also the reason why **2** is happening). I haven't quite figured out why yet.

**Comment by andysellick** [comment]
> 1. when I switch between tabs with VoiceOver, it reads out "Do you want to save X as your... etc" immediately after announcing the tab has been selected –  is this intended behaviour? https://user-images.githubusercontent.com/7116819/113029284-2527ad80-9184-11eb-8814-d1ec30144f39.mov

It's not intended behaviour. I think this might be the aria-live being triggered as a result of that element now being visible. As a general thing, I'm finding the tabs very difficult to use with Voiceover - selecting a tab jumps the focus down the page, the problem you described is inconsistent - which I think we might attribute to the tabs rather than this new functionality. Let's have a chat about this.

> 2. when I press the button to save a country, the state updates as expected, however when I try to undo my action, the button no longer seems to respond https://user-images.githubusercontent.com/7116819/113029687-9e270500-9184-11eb-985e-e290b6020f7e.mov

Yes - I had this bug and I fixed it. I think that the latest deploy to the preview app failed, so you're looking at a slightly older version. I've had this problem a lot with this PR, might be best to check out the branch locally for testing.

> 3. when I select a country as my default location, the UI state is updated but it is not maintained when I refresh the page. When I visit `/bank-holidays` without the hash, it shows me the default view. From poking around in dev tools, I can see it is not dropping the cookie as intended (which is probably also the reason why **2** is happening). I haven't quite figured out why yet.

If it's not saving the cookie at all, I might need to rebase this to get the latest components gem version, because that has a list of allowed cookies (I'll do that now).

**Review by danacotoran — APPROVED** [review]
If there is a decision to have this as a permanent feature after the experiment, it might be a good idea to go back and make this code more future proof. 

As it is right now, it does the job and I can't find any glaring issues so I am happy to approve.

File: app/views/calendar/_save_location.html.erb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+>
+  <h2 class="govuk-heading-m js-nation-description" aria-live="polite">
+    <%= sanitize(save_description) %>
+  </h2>
+
+  <%= render "govuk_publishing_components/components/button", {
+    text: save_button_text,
+    margin_bottom: 4,
+    classes: "js-nation-button",
+    aria_label: save_button_aria_label,
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/calendar/_save_location.html.erb:31]: I like the thinking behind that, but that would mean as they move from the button to the next element (the link) the text `We'll add a cookie to your device` would be duplicated. I agree this needs to be more accessible though.

**Pull request merged**